### ISSUE TITLE
Support notation switching

### ISSUE THREAD

**Issue by itsmonktastic** [EMGroup/js-eden#4]
If we want to support further TkEden notations then some work needs to be done to support notation switching, e.g:

```
%eden
x = 2;

%donald
line l
```

This could be done by generating some kind of "higher order" parser that looks for tokens like `%eden` and hand off everything else to other parsers stored in an associative array:

```
parsers[notationName](code);
```

**Comment by ElizabethHudnott** [comment]
While I have mixed feelings about multiple notations, I'd like to acknowledge that @itsmonktastic already put in the effort to make this work in master.

E.g.

%js
console.log("foo");

works!

There's also the beginnings of a DoNaLD implementation in there somewhere.  I don't know how complete it is though.

**Comment by knicos** [comment]
Granted, it exists but has it been used? More importantly, should it be used? Mixing notations is (I think) confusing and now that we are not so heavily based upon script files as before it seems much less relevant than it was in tkeden.

**Comment by ElizabethHudnott** [comment]
No, I don't think it has been used.  We could hardly encourage notation switching into JavaScript and DoNaLD was a work in progress that Meurig and Tim had on the go about New Year 2015 but then their respect attentions shifted elsewhere.

No doubt Meurig has many deep thoughts on this.

I had grand ideas back in 2007 that notations/domain specific languages were the future of computing and that domain experts (non-programmers) who already have paper based notations / specialized programming-like software would be able to construct things in our environment if only there was a general tool that could implement any industry-specific notation really quickly and that if this was the case then the harsh boundaries between the scribblings of domain experts and the writing of software specifications would merge together somewhat and that basically what tkeden was lacking was an effective meta-notation for rapidly prototyping notations.  Notations could be highly beneficial for collaboration with domain experts.  Implement one software specification and you have a rigid piece of software.  Implement the customer's native language and then they can tweak/change/add whatever features they want.  Dependencies and EDEN allow a lot of flexibility for changing things much more quickly than re-architecting a piece of conventional software but you still have to learn EDEN.

I still kind of believe that.  But on the other hand, having to learn SCOUT as well as EDEN just to draw a button on the screen proved to be something that slowed students' learning and made people less interested in EM.  When given a choice between using tkeden and JS-EDEN for their projects most of them chose JS-EDEN and then frequently cited the lack of notations as a reason for chosing JS-EDEN.  But then:

(1) SCOUT is a poor example of a notation.  There is no pre-established convention for notating user interfaces, as it's a topic that's internal to computer software.  And the "domain experts" of user interface authoring are often computer scientists themselves!  So basically it was just a another language (albeit a fancy one designed specifically for UI).  DoNaLD at least had links to learning Euclidian geometry in mathematics class, which was nice, but then math classes don't usually involve attributing colours, etc.  Also, mathematicians are often competent programmers anyway.  But a notation for some kind of important business application I could really see taking off, notwithstanding that often these pre-established industry conventions/notations can also be graphical as well textual.
(2) The domain of expertise of computer science students is almost by definition computer science.  Thus the people using these languages weren't domain experts at anything other than programming, so from a programmer's perspective why have 3 languages when one will do fine?

**Comment by knicos** [comment]
Perhaps, however, a more modern language for EM reduces the need to other notations also because it is less computer centric? Another option is to allow some flexibility within the plain eden language (custom operators?).

**Comment by ElizabethHudnott** [comment]
Yes, custom operators would be one valuable kind of flexibility that would enable to be more like the native language of the domain expert, but I was thinking more that parsing would be completely customizable.

Anyway, we probably have more than enough other things to think about right now.  It probably requires a proper and really quite broad literature review, etc.  No doubt also much work has already been done previously within the research group that I'm not aware of.

I think my feeling at the moment is that this isn't a priority.  If we succeed in getting a better EDEN language and some visual tools for making simple modifications without even needing to write code then we will have made significant progress.  I'm just not quite ready to say "never" when it comes to notations.  Also, while people still write words and numbers, I suspect that many worthy notations are graphical in nature also.

**Comment by knicos** [comment]
The AOP is probably the thing to look at re previous work in EM group. Its a notation for writing custom notations.

**Comment by ElizabethHudnott** [comment]
Yeah, the AOP is interesting.  Meurig sent me a link and I skimmed over it a while back (but not as long ago as 2007 when I was writing the PhD proposal and thinking about these things more serriously) :P
